### PR TITLE
refactor(neon_dashboard): Cache widgets and items

### PR DESCRIPTION
**Pull request by provokateurin**
TODO:
-  [x] Write tests for the whole bloc

### CODE REVIEW

File: packages/neon/neon_dashboard/lib/src/blocs/dashboard.dart [hunk 1]
```diff
@@ -16,112 +16,148 @@ sealed class DashboardBloc implements InteractiveBloc {
...
+      if (!result.hasData) {
+        return;
+      }
+
+      final v1WidgetIDs = ListBuilder<String>();
+      final v2WidgetIDs = ListBuilder<String>();
+
+      for (final widget in result.requireData) {
+        final itemApiVersions = widget.itemApiVersions;
+        if (itemApiVersions?.contains(2) ?? false) {
```
Comment (Leptopoda) [inline packages/neon/neon_dashboard/lib/src/blocs/dashboard.dart:43]:
```suggestion
        if (itemApiVersions != null && itemApiVersions.contains(2)) {
```

File: packages/neon/neon_dashboard/lib/src/blocs/dashboard.dart [hunk 2]
```diff
@@ -16,112 +16,148 @@ sealed class DashboardBloc implements InteractiveBloc {
...
-      }
+    await RequestManager.instance.wrapNextcloud(
+      account: account,
+      cacheKey: 'dashboard-widgets',
+      subject: widgets,
+      rawResponse: account.client.dashboard.dashboardApi.getWidgetsRaw(),
+      // Filter all widgets that don't support v1 nor v2
+      unwrap: (response) => response.body.ocs.data.values
+          .where((widget) => widget.itemApiVersions == null || widget.itemApiVersions!.isNotEmpty)
+          .toList(),
```
Comment (Leptopoda) [inline packages/neon/neon_dashboard/lib/src/blocs/dashboard.dart:113]: Can we make this a BuiltList please?

File: packages/neon/neon_dashboard/lib/src/blocs/dashboard.dart [hunk 3]
```diff
@@ -16,112 +16,148 @@ sealed class DashboardBloc implements InteractiveBloc {
...
 
-      this.widgets.add(
-            Result.success(
-              widgets.map(
-                (id, items) => MapEntry(
-                  response.body.ocs.data.values.firstWhere((widget) => widget.id == id),
-                  items,
+  void _updateItems() {
+    final resultV1 = itemsV1.valueOrNull;
+    final dataV1 = (resultV1?.hasData ?? false)
```
Comment (Leptopoda) [inline packages/neon/neon_dashboard/lib/src/blocs/dashboard.dart:119]:
```suggestion
    final dataV1 = (resultV1 != null && resultV1.hasData)
```

File: packages/neon/neon_dashboard/lib/src/blocs/dashboard.dart [hunk 4]
```diff
@@ -16,112 +16,148 @@ sealed class DashboardBloc implements InteractiveBloc {
...
-      debugPrint(s.toString());
-
-      widgets.add(Result.error(e));
-      return;
-    }
+          )
+        : null;
+
+    final resultV2 = itemsV2.valueOrNull;
+    final dataV2 = (resultV2?.hasData ?? false)
```
Comment (Leptopoda) [inline packages/neon/neon_dashboard/lib/src/blocs/dashboard.dart:136]:
```suggestion
    final dataV2 = (resultV2 != null && resultV2.hasData)
```

File: packages/neon/neon_dashboard/lib/src/blocs/dashboard.dart [hunk 5]
```diff
@@ -16,112 +16,148 @@ sealed class DashboardBloc implements InteractiveBloc {
...
+  void _updateItems() {
+    final resultV1 = itemsV1.valueOrNull;
+    final dataV1 = (resultV1?.hasData ?? false)
+        ? Map.fromEntries(
+            resultV1!.requireData.entries.map(
+              (e) => MapEntry(
+                e.key,
+                dashboard.WidgetItems(
+                  (b) => b
+                    ..items = e.value.sublist(0, min(e.value.length, maxItems)).toBuilder()
```
Comment (Leptopoda) [inline packages/neon/neon_dashboard/lib/src/blocs/dashboard.dart:126]:
```suggestion
                    ..items.replace(e.value.sublist(0, min(e.value.length, maxItems)))
```

File: packages/neon/neon_dashboard/lib/src/blocs/dashboard.dart [hunk 6]
```diff
@@ -16,112 +16,148 @@ sealed class DashboardBloc implements InteractiveBloc {
...
+      Result(
+        <String, dashboard.WidgetItems>{
+          ...?dataV1,
+          ...?dataV2,
+        },
+        resultV1?.error ?? resultV2?.error,
+        isLoading: (resultV1?.isLoading ?? true) || (resultV2?.isLoading ?? true),
+        isCached: (resultV1?.isCached ?? true) || (resultV2?.isCached ?? true),
+      ),
+    );
```
Comment (Leptopoda) [inline packages/neon/neon_dashboard/lib/src/blocs/dashboard.dart:161]: What about creating one map directly?
```dart
    final items = MapBuilder<String, dashboard.WidgetItems>();

    final resultV1 = itemsV1.valueOrNull;
    if (resultV1 != null && resultV1.hasData) {
      for (final entry in resultV1.requireData.entries) {
        items[entry.key] = dashboard.WidgetItems(
          (b) => b
            ..items.replace(entry.value.sublist(0, min(entry.value.length, maxItems)))
            ..emptyContentMessage = ''
            ..halfEmptyContentMessage = '',
        );
      }
    }

    final resultV2 = itemsV2.valueOrNull;
    if (resultV2 != null && resultV2.hasData) {
      for (final entry in resultV2.requireData.entries) {
        items[entry.key] = entry.value.rebuild((b) {
          if (b.items.length > maxItems) {
            b.items.removeRange(maxItems, b.items.length);
          }
        });
      }
    }

    this.items.add(
          Result(
            items.build(),
            resultV1?.error ?? resultV2?.error,
            isLoading: (resultV1?.isLoading ?? true) || (resultV2?.isLoading ?? true),
            isCached: (resultV1?.isCached ?? true) || (resultV2?.isCached ?? true),
          ),
        );
```

**Review by Leptopoda — COMMENTED** [review]
Some minor nitpicks

**Comment by provokateurin** [comment]
@Leptopoda tests are done, ready to be reviewed

**Review by Leptopoda — APPROVED** [review]

**Pull request merged**